cmd/ko/README.md

```diff
@@ -0,0 +1,146 @@
+# `ko`
+
+## What is `ko`?
+
+`ko` is an **experimental** CLI to support rapid development of Go containers
+with Kubernetes.
+
+## Installation
+
+`ko` can be installed via:
+
+```shell
+go install github.com/google/go-containerregistry/cmd/ko
+```
+
+## The `ko` Model
+
+`ko` is built around a very simple extension to Go's model for expressing
+dependencies using import paths. In Go, dependencies are expressed via
+blocks like:
+
+```go
+import (
+    "github.com/google/go-containerregistry/authn"
+    "github.com/google/go-containerregistry/name"
+)
+```
+
+Similarly (as you can see above), Go binaries can be references via import
+paths like `github.com/google/go-containerregistry/cmd/ko`.
+
+With `ko`, we simply replace image references within Kubernetes yaml
+files with the import paths to the Go binaries that should be containerized
+and substituted there. For example, you might use the following in a
+Kubernetes `Deployment` resource:
+
+```yaml
+apiVersion: apps/v1beta1
+kind: Deployment
+metadata:
+  name: hello-world
+spec:
+  replicas: 1
+  template:
+    spec:
+      containers:
+      - name: hello-world
+        # This is the importpath for the Go binary to build and run.
+        image: github.com/examples/http/cmd/helloworld
```

Comment:
It feels weird to me that we're asking users to write a k8s-style YAML that won't be valid until you've `ko resolve`d it. Maybe we can signal that the YAML isn't really ready to use yet, by naming it `deployment.yaml.tmpl` or something in this example? Or otherwise making it clear that the `github.com/whatever/cmd/hello` is a placeholder which will be resolved? Possibly by surrounding it with `{{}}` ? Does Kubernetes have any kind of standard for this today?

Reply:
To your point I wonder if instead of hijacking the `image` property you use `goSource` or something as explicit

also `deployment.yaml.tmpl` -> `deployment.yaml.ko` ?

Reply:
**tl;dr** In general, I have a bias towards adopting facets of `rules_k8s` that seem to be working well.

More and more in the context of Bazel, I've biased towards using non-sensical image names that are given meaning by `image_chroot` ([e.g.](https://github.com/mattmoor/warm-image/blob/master/controller.yaml#L16)), and this is effectively the same idea.  Since skaffold came up, I'll also point out that [this](https://github.com/GoogleContainerTools/skaffold/blob/master/examples/getting-started/skaffold.yaml#L5) never publishes a `:latest` tag, which is what is expected in all of the yamls :)

re: standard substitutions, no.  Part of my concern with `{{}}` is that it's invalid yaml, which introduces new fun.

re: `image:`, this isn't the only context for substitutions.  For example, it is an extremely useful pattern to be able to pass resolved container names as flags to controllers ([e.g.](https://github.com/mattmoor/warm-image/blob/master/controller.yaml#L21)).  Also, I find there is value in avoiding schema-sensitivity, since it means this model should work with most CRDs that deal with images as well ([e.g.](https://github.com/mattmoor/warm-image/blob/master/example-warmimage.yaml#L6)).

re: file extension, K8s doesn't have a monopoly on `yaml`.  :)

---
Ultimately, I am really trying to minimize the number of concepts folks need to grok to use this and everything we add increases the learning curve.

Reply:
I understand the motivation, and certainly introducing a new concept early on might confuse users.

My concern is that by piggybacking on Go import paths, users won't realize that their image is actually at `gcr.io/my-project/github.com/user/repo/cmd/foo/hello`, and not knowing that linkage will make it harder for them to debug issues in the future, for instance when they try to pull the image that's acting funny in the cluster to their local workstation.

A magical onboarding can be harmful if you don't understand the underlying magic. I think of it like the Sorcerer's Apprentice, you want to know at least a little about what's happening if it ever goes wrong. Maybe that's not the case for Bazel today either, I don't know.

It's probably more a documentation issue. We can spell it out extra clearly in the README that `github.com/foo/...` actually means `$KO_DOCKER_REPO/github.com/foo/...` and repeat it three times before you proceed. :grinning:

Reply:
Perhaps it would help to share the output of a `ko apply`, since I think it helps clarify.

I'm modifying WarmImage to use this, and here's the output when I apply:
```shell
$ ko apply -f config/
2018/04/25 12:14:41 Go building github.com/mattmoor/warm-image/cmd/sleeper
2018/04/25 12:14:41 Go building github.com/mattmoor/warm-image/cmd/controller
2018/04/25 12:14:43 Publishing us.gcr.io/convoy-adapter/github.com/mattmoor/warm-image/cmd/sleeper:latest
2018/04/25 12:14:43 mounted sha256:eb05f3dbdb543cc610527248690575bacbbcebabe6ecf665b189cf18b541e3ca
2018/04/25 12:14:43 mounted sha256:3ca7d60fa89dc8a1faea3046fd3516f23dab93489f3888ae539df4abc0973e52
2018/04/25 12:14:43 mounted sha256:e2cc7c829942a768015dbcfdad7205c104cb85b84a79573555a8f4381b98110c
2018/04/25 12:14:44 pushed us.gcr.io/convoy-adapter/github.com/mattmoor/warm-image/cmd/sleeper:latest
2018/04/25 12:14:44 Published us.gcr.io/convoy-adapter/github.com/mattmoor/warm-image/cmd/sleeper@sha256:193acdbeff1ea9f105f49d97a6ceb7adbd30b3d64a8b9949382f4be9569cd06d
2018/04/25 12:14:50 Publishing us.gcr.io/convoy-adapter/github.com/mattmoor/warm-image/cmd/controller:latest
2018/04/25 12:14:51 mounted sha256:fbc44e14a1d848ed485b5c3f03611c3e21aaa197fcba579255e5f8416a1b7172
2018/04/25 12:14:51 mounted sha256:eb05f3dbdb543cc610527248690575bacbbcebabe6ecf665b189cf18b541e3ca
2018/04/25 12:14:51 mounted sha256:dc0dd55edef1443e976c835825479c3dc713bb689547f8a170f8a0d14f9ff734
2018/04/25 12:14:52 pushed us.gcr.io/convoy-adapter/github.com/mattmoor/warm-image/cmd/controller:latest
2018/04/25 12:14:52 Published us.gcr.io/convoy-adapter/github.com/mattmoor/warm-image/cmd/controller@sha256:78794915fca48d0c4b339dc1df91a72f1e4bc6a7b33beaeef8aecda0947d5d31
clusterrolebinding "warmimage-controller-admin" configured
deployment "warmimage-controller" created
namespace "warmimage-system" configured
serviceaccount "warmimage-controller" unchanged
customresourcedefinition "warmimages.mattmoor.io" configured
```

I think that this clearly indicates that the images are being published as `$KO_DOCKER_REPO/github.com/foo/...`, but I could be wrong?  WDYT?

Reply:
Re: hijacking the `image` property. Using a stand-in image reference here feels dirty to me. I'm willing to be convinced that I'm wrong, but IMO this would be cleaner if you used another key, e.g. "import":

```yaml
    spec:
      containers:
      - name: hello-world
        # This is the import path for the Go binary to build and run.
        import: github.com/mattmoor/examples/http/cmd/helloworld
        ports:
        - containerPort: 8080
```

It's very clear that is *not* an image reference, and `import` would be replaced by an equivalent `image` by `ko`.

Reply:
To reiterate:
> re: image:, this isn't the only context for substitutions. For example, it is an extremely useful pattern to be able to pass resolved container names as flags to controllers (e.g.). Also, I find there is value in avoiding schema-sensitivity, since it means this model should work with most CRDs that deal with images as well (e.g.).

At the end of the day, this is an experimental tool.  I see this as the heart of that experiment, and I'm pretty biased against changing it before even trying it.